### DISCUSSION TITLE
BepuPhysic - Materials, Friction, Bounciness

### DISCUSSION THREAD

**Discussion by Magic73** [bepu/bepuphysics2#236] [Q&A]
It has been a long time since I last used Bepu. Yesterday I was experimenting with Unity but its physics have limitations for what I was trying to do. I remembered BEPU, and saw if it was possible to use it within Unity. I found the AntoineCharton repository (with version 2.1 support), and tried using it in Unity.
It worked, and I updated the version to 2.2 (for 2.3.4 I had the problem of the Stepper that I have not seen / read how to configure it).

Anyway, I managed to get my test project to work:
pool table, rails, a ball and cue.
And the result alone in this way is much better than the Unity counterpart.

The problem is that now I should add the characteristics to the ball, the table and the sides, in terms of friction (static / dynamic) and rebound.

I tried to see the demo code, but I got lost. several things have changed since I used BEPU years ago.

after creating an object, how do i assign these properties to it?

**Comment by RossNordby** [comment]
In bepuphysics2, there is no built-in 'material' for individual bodies or collidables. Instead, it's handled through the `INarrowPhaseCallbacks`.

I'd recommend reading through these:
https://github.com/bepu/bepuphysics2/blob/master/Documentation/GettingStarted.md
https://github.com/bepu/bepuphysics2/blob/master/Demos/Demos/FrictionDemo.cs
https://github.com/bepu/bepuphysics2/blob/master/Demos/Demos/BouncinessDemo.cs

(Notably, while Unity is making progress and bepuphysics2 apparently works better in it than it used to, you should still expect a major performance penalty until it finishes some [runtime work](https://forum.unity.com/threads/unity-future-net-development-status.1092205/page-23)- or you jump through several hoops to integrate it as a [native ](https://github.com/RossNordby/scratchpad/tree/main/Abomination/AbominationInterop) plugin.)

**Comment by Magic73** [comment]
Hi @RossNordby  :)
I've some question about the setup of the narrowphase.
FrictionDemo.cs:

```
                collidableMaterials.Allocate(Simulation.Bodies.Add(boxDescription)) = new SimpleMaterial
                {
                    FrictionCoefficient = maximumFriction * i / (boxCount - 1f),
                    MaximumRecoveryVelocity = 2,
                    SpringSettings = new SpringSettings(30, 1)
                };
```

with these lines I register the "material" properties to the box, so that it can be retrived inside the callback?


as for the parameters, what are the range of values ​​in which to move?
I mean for the bounce, what range of values ​​should I use to have "no bounce" and "perfect bounce"?
for friction, what values ​​should I use to have ice effect, and maximum resistance?
(as far as friction is concerned, in Unity it is divided between dynamic friction and static friction.How can I set the latter?)


In the simulation I may use one single narrowphasecallback, but then why in your demos there are multiple definitions of them?

**Reply by RossNordby** [reply]
> with these lines I register the "material" properties to the box, so that it can be retrived inside the callback?

Basically, yup. Worth noting that there's nothing special about the `CollidableProperty` used in the demos; it's just a convenience type. Nothing more than an array indexed by handles. You don't have to use it if you don't want to.

> as for the parameters, what are the range of values ​​in which to move?

Friction is any nonnegative number. Maximum friction force is the normal force scaled by friction. 0 is perfectly slippery, higher numbers are higher friction. There is no static friction at the moment.

To make something bounce, you'll want `MaximumRecoveryVelocity` at a high value (possibly float.MaxValue), a `SpringSettings` damping ratio of 0 (that is, no damping), and a `SpringSettings` frequency low enough that your simulation can represent the bounce motion. Substepping can help there. See the BouncinessDemo for more details.

For something that's not bouncy, using a lower `MaximumRecoveryVelocity` (like 2) is a good choice. At a 60hz update rate, 30hz spring frequency and 1 damping ratio is a good default for most stuff. A damping ratio of 1 means the contact springs are critically damped. Higher values are overdamped. See wikipedia for more: https://en.wikipedia.org/wiki/Damping

> In the simulation I may use one single narrowphasecallback, but then why in your demos there are multiple definitions of them?

The demos are designed to show different simulation configurations. Having every possible demo stuffed into one callbacks type would be pretty bad for learning about a particular demo that uses only a tiny fraction of it.

**Reply by Magic73** [reply]
using values ​​like MaxValue doesn't make it easy to understand what the right compromise might be.

for example, losing 50% of the speed of a bounce, how do I determine that?

excluding the desire to try. :)

**Reply by RossNordby** [reply]
There's not a perfect mapping from coefficient of restitution to spring settings. The contact spring settings act like an actual spring, and the motion takes place over multiple timesteps (or substeps). A damping ratio of 0 damps the motion as little as is possible for the integrator to represent, a damping ratio of 1 returns the spring to 0 compression as quickly as possible without overshooting (within numerical error). The closer you get to 0, the less damped the motion is, and the more it will bounce. Going above 1 makes it take longer to return to 0 compression without bouncing.

Assuming no integrator damping losses, the curve relating coefficient of restitution to damping ratios looks something like: https://www.researchgate.net/figure/Relation-between-impact-damping-ratio-and-coefficient-of-restitution-CR-according-to_fig2_335073194
